src/BlockDiagonals.jl

```diff
@@ -0,0 +1,338 @@
+module BlockDiagonals
+
+using BlockArrays
+using BlockArrays: AbstractBlockSizes, BlockSizes
+using FillArrays
+using LinearAlgebra
+
+export BlockDiagonal, blocks
+# reexport core interfaces from BlockArrays
+export Block, BlockSizes, blocksize, blocksizes, nblocks
+
+"""
+    BlockDiagonal{T, V<:AbstractMatrix{T}, S<:AbstractBlockSizes} <: AbstractBlockMatrix{T}
+
+A BlockMatrix with square blocks of type `V` on the diagonal, and zeros off the diagonal.
+"""
+struct BlockDiagonal{T, V<:AbstractMatrix{T}, S<:AbstractBlockSizes} <: AbstractBlockMatrix{T}
+    blocks::Vector{V}
+    blocksizes::S
+
+    function BlockDiagonal{T, V, S}(blocks::Vector{V}, sizes::S) where {T, V<:AbstractMatrix{T}, S<:AbstractBlockSizes}
+
+        all(is_square, blocks) || throw(ArgumentError("All blocks must be square."))
+        return new{T, V, S}(blocks, sizes)
+    end
+end
+
+function BlockDiagonal(blocks::Vector{V}, sizes::S) where {T, V<:AbstractMatrix{T}, S<:AbstractBlockSizes}
+    return BlockDiagonal{T, V, S}(blocks, sizes)
+end
+
+function BlockDiagonal(blocks::AbstractVector{<:AbstractMatrix})
+    # mapreduce would give Tuples; need Arrays
+    sizes = BlockSizes(([size(b, 1), size(b, 2)] for b in blocks)...)
+    return BlockDiagonal(blocks, sizes)
+end
+
+BlockDiagonal(B::BlockDiagonal) = copy(B)
+
+is_square(A::AbstractMatrix) = size(A, 1) == size(A, 2)
+
+"""
+    blocks(B::BlockDiagonal{T, V}) -> Vector{V}
+
+Return the on-diagonal blocks of B.
+"""
+blocks(B::BlockDiagonal) = B.blocks
+
+## AbstractBlockMatrix interface
+BlockArrays.blocksizes(B::BlockDiagonal) = B.blocksizes
+
+# Needs to be `Int` not `Integer` to avoid methods ambiguity. Can be changed after
+# BlockArrays v0.9 is released; see https://github.com/JuliaArrays/BlockArrays.jl/issues/82
+BlockArrays.blocksize(B::BlockDiagonal, p::Int)::Tuple = size(blocks(B)[p])
+function BlockArrays.blocksize(B::BlockDiagonal, p::Int, q::Int)::Tuple
+    return size(blocks(B)[p], 1), size(blocks(B)[q], 2)
+end
+
+# Needs to be `Int` not `Integer` to avoid methods ambiguity. Can be changed after
+# BlockArrays v0.9 is released; see https://github.com/JuliaArrays/BlockArrays.jl/issues/82
+BlockArrays.nblocks(B::BlockDiagonal, dim::Int) = dim > 2 ? 1 : length(blocks(B))
+BlockArrays.nblocks(B::BlockDiagonal) = length(blocks(B)), length(blocks(B))
+
+# Needs to be `Int` not `Integer` to avoid methods ambiguity. Can be changed after
+# BlockArrays v0.9 is released; see https://github.com/JuliaArrays/BlockArrays.jl/issues/82
+function BlockArrays.getblock(B::BlockDiagonal{T}, p::Int, q::Int) where T
+    return p == q ? blocks(B)[p] : Zeros{T}(blocksize(B, p, q))
```

Comment:
This is the only use of FillArrays

This function is also not type-stable... (and I don't think it can be? becuase the diagonal blocks can be any AbstractMatrix, e.g., Diagonal, which doesn't really make sense for all-zero off-diagonal blocks)

Reply:
I think this is fine, for now